tests/test_mutations.py

```diff
@@ -472,47 +705,288 @@ def test_keep_mutation_parent_zero_rate(self):
-    # Insert a sentinel into the map for convenience.
-    map_position = np.hstack([mutation_map.position, [tables.sequence_length]])
-    rng = msprime.RandomGenerator(seed)
-    time = tables.nodes.time
-    for edge in tables.edges:
-        branch_length = time[edge.parent] - time[edge.child]
-        index = np.searchsorted(map_position, edge.left)
-        if map_position[index] > edge.left:
-            index -= 1
-        left = edge.left
-        right = 0
-        while right != edge.right:
-            right = min(edge.right, map_position[index + 1])
-            assert left < right
-            assert mutation_map.position[index] <= left
-            assert right <= map_position[index + 1]
-            assert right <= edge.right
-            # Generate the mutations.
-            mu = mutation_map.rate[index] * (right - left) * branch_length
-            for _ in range(rng.poisson(mu)):
-                position = rng.flat(left, right)
-                site_id = tables.sites.add_row(position, ancestral_state="0")
-                tables.mutations.add_row(site_id, node=edge.child, derived_state="1")
-                # Sample from this to duplicate the behaviour in the C implementation.
-                rng.uniform_int(1)
-            index += 1
-            left = right
-    tables.sort()
-    return tables.tree_sequence()
+
+    def __init__(self, rate_map, alphabet=None,
+                 mutation_matrix=None, alphabet_distribution=None):
+        """
+        Defaults to all 0->1 mutations.
+        """
+        if alphabet is None:
+            alphabet = msprime.BINARY
+        if mutation_matrix is None:
+            mutation_matrix = np.array([[0.0, 1.0], [1.0, 0.0]])
+        if alphabet_distribution is None:
+            alphabet_distribution = np.array([1.0, 0.0])
+        self.rate_map = rate_map
+        self.alphabet = alphabet
+        self.mutation_matrix = mutation_matrix
+        self.alphabet_distribution = alphabet_distribution
+        self.alleles = {msprime.BINARY: tskit.ALLELES_01,
+                        msprime.NUCLEOTIDES: tskit.ALLELES_ACGT}[self.alphabet]
+        self.sites = []
+
+    def print_state(self):
+        print("alleles", self.alleles)
+        for site in self.sites:
+            print(site)
+
+    def generate(self, tables, seed, keep=False, discrete_sites=False):
+        rng = msprime.RandomGenerator(seed)
+        if keep:
+            self.initialise_sites(tables)
+        tables.sites.clear()
+        tables.mutations.clear()
+        self.place_mutations(tables, rng, discrete_sites=discrete_sites)
+        self.choose_alleles(tables, rng)
+        self.populate_tables(tables)
+        self.record_provenance(tables, seed, keep, discrete_sites)
+        return tables.tree_sequence()
+
+    def initialise_sites(self, tables):
+        sites = tables.sites
+        mutations = tables.mutations
+        nodes = tables.nodes
+        j = 0
+        for site_id in range(sites.num_rows):
```

Comment:
I would structure this as 
```python
mutation_rows = iter(tables.mutations)
mutation_row = next(mutation_rows, None)
for site_row in tables.sites:
     site_mutations = []
     while mutation_row is not None and mutation_row.site == site_row.id:
           mutation = Mutation(...)
           site_mutations.append(mutation)
           mutation_row = next(mutation_rows, None)
     self.sites[site_row.position] = Site(...)
```

It's the same thing, I know, but I think it's really important to get this code as simple and clear as possible so that people can understand the algorithm as quickly as possible.

Reply:
Much clearer, thanks.